### ISSUE TITLE
[Feature Request] Include/exclude obj from decompression matching regex

### ISSUE THREAD

**Issue by st-gr** [qpdf/qpdf#404]
Hi Jay,

First and foremost, I would like to thank you for your great tool that I use on 100k of pdf’s per day.

Recently I came up with a **requirement**, where I only would want to de-compress/filter certain objects, e. g. containing BT..ET, but no /Image or /FontFile using either the --qdf option or --stream-data=uncompress.  There is already a filter for certain encodings, but I would like to filter the in-memory uncompressed objects for regex like “/BT\s.*\sET/” and then keep all other objects compressed.

The new filter should have two options: inclusive or exclusive.  This way one could ex- or include certain obj contents to be de-compressed in the output providing a JSON of 1 or multiple regex. 
E. g. 

> `--decompress-filter={"include":["BT\\s.*\\sET"],"exclude":["^\\s*/FontFile.{1,2}(\\d+)  \\d+ .*","^\\s+/Image\\d+ (\\d+) \\d+ .*"]}`

The first group in the regex would match the obj to be included or excluded.  If no group is defined then one match would qualify the surrounding obj to be in- or excluded.

I tried to create a new version of the c++ **fix-qdf** tool to copy certain objects from an uncompressed version of the pdf back into the QDF document in process.  However, it looks like that the fix-qdf parser only supports QDF formatted pdf files and not any other regular pdf’s.

Is this a feature that you would deem feasible for a future release or could you point me to a workaround or qpdf source code positions I would need to look at to enhance?

Thank you,
Stefan

**Comment by jberkenbilt** [comment]
Hello. I'm glad qpdf is useful to you!

I'm not sure I think a feature like the one your suggesting would be useful for the CLI tool. I think when you get to that level of specificity, you are better off coding against the library API. Using the qpdf library, it would certainly be possible to implement something like what you're suggesting. That said, I am definitely open to a discussion of how we could enhance the CLI to make it possible to implement something like what you're trying to achieve.

Presumably if you're processing that many files with complicated requirements, you must be doing something programmatically. What language or languages are you using? I'm not sure if you're aware of [pikepdf](https://pypi.org/project/pikepdf/), which is a set of Python bindings for the qpdf library. I am not the author, but I use it and think it's of very high quality. It might be possible for you to use it to do what you're trying to do. Perhaps there may be some library enhancements I could make that would improve the ease of doing this kind of operation, but I haven't thought it through at this time.

Anyway, if you let me know how you're using qpdf, maybe I can help think of a good solution.

**Comment by jberkenbilt** [comment]
If you are using the C++ library, you can use `QPDFWriter::setCompressStreams(false)` to prevent QPDFWriter from compressing streams automatically when it writes. Streams that are already compressed will stay compressed. For streams that you want to remain uncompressed, you can explicitly replace their stream data in the original QPDF object with the uncompressed data from the stream. For example, you can use `QPDF::replaceStreamData` with the `PointerHolder<Buffer>` that was returned by `QPDF::getStreamData` on the same stream, thus replacing the compressed stream with an uncompressed stream. Writing some code that walks through streams applying matching logic and then operates in this way could probably get you what you want, and I suspect it would be pretty easy to write this in Python with pikepdf, or it could be written in C++. I'm happy to discuss further. I may work on other language bindings to qpdf in the future, but I don't have any immediate plans to do so right now.

**Comment by st-gr** [comment]
Hi Jay,

Thank you for your swift and helpful reply.

The application I use your tool for is long time archival of PDF, having their texts in pseudo readable form, but not in a separate file (that would be easy to achieve).

A subsequent tool I need to use in the process runs in `O((i+k) log i)` time for i/k = different file sizes.  In my case a regular compressed pdf file of 200 KB becomes 36 MB.  That is why I would rather not want to use the completely uncompressed versions of the pdf to feed into the above tool, but rather partial uncompressed to optimize the runtime according to Big O time-dependency above.

I am still using the CLI version of QPDF, but massive parallel on multiple nodes on a hyperscaler.
Since I must process this many pdf only one time (later incremental archival) I thought I could get by using the CLI version alone.

I am not shy in implementing the partial decompression in C++ applying the tips you gave me.  At first it surely will be faster to implement it in Python though as you pointed out.
I am sorry, I am new to the QPDF library.  Would the `pdf-invert-images.cc` example be a good template to start with?  The fix-qdf tool uses `QUtil::read_lines_from_file` which returns a nice vector of lines which are easy to match with regex.  I just wonder how would I identify the obj in a stream?

When I think about the above algorithm then testing the regex for each line will also lead to a performance penalty.  Oh well, there is no free lunch.  I could take shortcuts and code specific to my internal pdf file structure, but that will make it not flexible and thus unusable for other pdf files.

I am glad I reached out to you.  In the past I always tried to figure out things on my own which I still consider the best way to learn, but I simply don’t have the time any more to do the research myself.  I miss the old days writing TSR programs for fun in Assembler on M$ DOS without the ability to Google/Altavista code snippets.

Thanks again for the hints and support you provided.

Thank you,
Stefan

**Comment by jberkenbilt** [comment]
pdf-invert-images.cc would be a reasonable starting point. If I have some time in the next few days, I could probably produce some example code. There were some recent additions to qpdf to allow you to get information about where in a file certain things appear. The code was contributed and first appeared in 9.1.0. Anyway, my hunch is that it shouldn't take much code to get this to work. I'm happy to look over any code you've written or offer suggestions along the way.

We must have started programming at around the same time. I wrote my first programs on graph paper, but the first ones I actually ran on a computer were on a TRS-80 circa 1981. I had already been programming for 13 years before I had access to the web. I had to learn stuff by reading books, reading source code, and figuring it out. :-)
